File: app/src/main/java/com/zxcx/zhizhe/ui/loginAndRegister/login/LoginActivity.java
package com.zxcx.zhizhe.ui.loginAndRegister.login;

import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextPaint;
import android.text.TextWatcher;
import android.view.KeyEvent;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

import com.meituan.android.walle.WalleChannelReader;
import com.zxcx.zhizhe.R;
import com.zxcx.zhizhe.event.LoginEvent;
import com.zxcx.zhizhe.event.RegisterEvent;
import com.zxcx.zhizhe.mvpBase.MvpActivity;
import com.zxcx.zhizhe.ui.loginAndRegister.channelRegister.ChannelRegisterActivity;
import com.zxcx.zhizhe.ui.loginAndRegister.forget.ForgetPasswordActivity;
import com.zxcx.zhizhe.ui.loginAndRegister.register.RegisterActivity;
import com.zxcx.zhizhe.utils.Constants;
import com.zxcx.zhizhe.utils.MD5Utils;
import com.zxcx.zhizhe.utils.Utils;
import com.zxcx.zhizhe.utils.ZhiZheUtils;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.HashMap;
import java.util.regex.Pattern;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import cn.jiguang.analytics.android.api.JAnalyticsInterface;
import cn.sharesdk.framework.Platform;
import cn.sharesdk.framework.PlatformActionListener;
import cn.sharesdk.framework.PlatformDb;
import cn.sharesdk.framework.ShareSDK;
import cn.sharesdk.sina.weibo.SinaWeibo;
import cn.sharesdk.tencent.qq.QQ;
import cn.sharesdk.wechat.friends.Wechat;

public class LoginActivity extends MvpActivity<LoginPresenter> implements LoginContract.View {

    @BindView(R.id.et_login_phone)
    EditText mEtLoginPhone;
    @BindView(R.id.et_login_password)
    EditText mEtLoginPassword;
    @BindView(R.id.btn_login)
    Button mBtnLogin;
    @BindView(R.id.iv_login_phone_clear)
    ImageView mIvLoginPhoneClear;

    String phoneRules = "^1\\d{10}$";
    String passwordRules = "^[a-zA-Z0-9]{6,16}$";
    Pattern phonePattern = Pattern.compile(phoneRules);
    Pattern passwordPattern = Pattern.compile(passwordRules);

    PlatformActionListener mChannelLoginListener = new ChannelLoginListener();
    private int channelType; // 1-QQ 2-WeChat 3-Weibo
    private int appType;
    private String userName, userId, userIcon, userGender, appChannel, appVersion;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);
        ButterKnife.bind(this);
        EventBus.getDefault().register(this);
        initView();

        appType = Constants.APP_TYPE;
        appChannel = WalleChannelReader.getChannel(this);
        appVersion = Utils.getAppVersionName(this);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        EventBus.getDefault().unregister(this);
    }

    private void initView() {
        mEtLoginPhone.addTextChangedListener(new LoginTextWatcher());
        mEtLoginPhone.addTextChangedListener(new PhoneTextWatcher());
        mEtLoginPassword.addTextChangedListener(new LoginTextWatcher());
        mEtLoginPassword.setOnEditorActionListener(new LoginListener());
        TextPaint paint = mBtnLogin.getPaint();
        paint.setFakeBoldText(true);
    }

    @Override
    protected LoginPresenter createPresenter() {
        return new LoginPresenter(this);
    }

    @Override
    public void getDataSuccess(LoginBean bean) {
        ZhiZheUtils.saveLoginData(bean);
        //极光统计
        cn.jiguang.analytics.android.api.LoginEvent lEvent = new cn.jiguang.analytics.android.api.LoginEvent("defult", true);
        lEvent.addKeyValue("appChannel", WalleChannelReader.getChannel(this)).addKeyValue("appVersion", Utils.getAppVersionName(this));
        JAnalyticsInterface.onEvent(this, lEvent);
        //登录成功通知
        EventBus.getDefault().post(new LoginEvent());
        EventBus.getDefault().postSticky(new LoginEvent());
        Utils.closeInputMethod(mEtLoginPassword);
        finish();
    }

    @Override
    public void channelLoginSuccess(LoginBean bean) {
        getDataSuccess(bean);
    }

    @Override
    public void channelLoginNeedRegister() {
        toastShow("请绑定手机号码");
        Intent intent = new Intent(this, ChannelRegisterActivity.class);
        intent.putExtra("userId",userId);
        intent.putExtra("userName",userName);
        intent.putExtra("userIcon",userIcon);
        intent.putExtra("userGender",userGender);
        intent.putExtra("channelType",channelType);
        startActivity(intent);
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onMessageEvent(RegisterEvent event) {
        //登录成功通知
        EventBus.getDefault().post(new LoginEvent());
        EventBus.getDefault().postSticky(new LoginEvent());
        finish();
    }

    @OnClick(R.id.iv_login_phone_clear)
    public void onMIvLoginPhoneClearClicked() {
        mEtLoginPhone.setText("");
    }

    @OnClick(R.id.tv_login_register)
    public void onMTvLoginRegisterClicked() {
        Intent intent = new Intent(this, RegisterActivity.class);
        startActivity(intent);
    }

    @OnClick(R.id.tv_login_forget_password)
    public void onMTvLoginForgetPasswordClicked() {
        Intent intent = new Intent(this, ForgetPasswordActivity.class);
        startActivity(intent);
    }

    @OnClick(R.id.btn_login)
    public void onMBtnLoginClicked() {
        login();
    }

    @OnClick(R.id.ll_login_qq)
    public void onMLlLoginQqClicked() {
        channelType = 1;
        Platform qq = ShareSDK.getPlatform(QQ.NAME);
        qq.setPlatformActionListener(mChannelLoginListener);
        qq.showUser(null);
    }

    @OnClick(R.id.ll_login_wechat)
    public void onMLlLoginWechatClicked() {
        channelType = 2;
        Platform wechat = ShareSDK.getPlatform(Wechat.NAME);
        wechat.setPlatformActionListener(mChannelLoginListener);
        wechat.showUser(null);
    }

    @OnClick(R.id.ll_login_weibo)
    public void onMLlLoginWeiboClicked() {
        channelType = 3;
        Platform weibo = ShareSDK.getPlatform(SinaWeibo.NAME);
        weibo.setPlatformActionListener(mChannelLoginListener);
        weibo.showUser(null);
    }

    @OnClick(R.id.iv_login_close)
    public void onViewClicked() {
        finish();
    }

    private void login() {
        if (checkPhone() && checkPassword()) {
            String phone = mEtLoginPhone.getText().toString();
            String password = MD5Utils.md5(mEtLoginPassword.getText().toString());
            mPresenter.phoneLogin(phone, password, appType, appChannel, appVersion);
        }
    }

    private boolean checkPhone() {
        if (phonePattern.matcher(mEtLoginPhone.getText().toString()).matches()) {
            return true;
        } else {
            toastShow("手机号格式错误!");
            return false;
        }
    }

    private boolean checkPassword() {
        if (passwordPattern.matcher(mEtLoginPassword.getText().toString()).matches()) {
            return true;
        } else {
            toastShow("密码格式错误!");
            return false;
        }
    }

    class ChannelLoginListener implements PlatformActionListener {

        @Override
        public void onComplete(Platform platform, int i, HashMap<String, Object> hashMap) {
            if (i == Platform.ACTION_USER_INFOR) {
                PlatformDb platDB = platform.getDb();//获取数平台数据DB
                //通过DB获取各种数据
                userGender = platDB.getUserGender();
                userIcon = platDB.getUserIcon();
                userId = platDB.getUserId();
                userName = platDB.getUserName();
                mPresenter.channelLogin(channelType,userId,appType,appChannel,appVersion);
            }
        }

        @Override
        public void onError(Platform platform, int i, Throwable throwable) {
            throwable.printStackTrace();
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    toastShow("登录失败");
                }
            });
        }

        @Override
        public void onCancel(Platform platform, int i) {
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    toastShow("登录取消");
                }
            });
        }
    }

    class LoginTextWatcher implements TextWatcher {

        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {

        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {

        }

        @Override
        public void afterTextChanged(Editable s) {
            if (mEtLoginPhone.length() > 0 && mEtLoginPassword.length() > 0) {
                mBtnLogin.setEnabled(true);
            } else {
                mBtnLogin.setEnabled(false);
            }
        }
    }


    class PhoneTextWatcher implements TextWatcher {

        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {

        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {

        }

        @Override
        public void afterTextChanged(Editable s) {
            if (mEtLoginPhone.length() > 0) {
                mIvLoginPhoneClear.setVisibility(View.VISIBLE);
            } else {
                mIvLoginPhoneClear.setVisibility(View.GONE);
            }
        }
    }

    private class LoginListener implements TextView.OnEditorActionListener {
        @Override
        public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
            //此处会响应2次 分别为ACTION_DOWN和ACTION_UP
            if (actionId == EditorInfo.IME_ACTION_SEARCH
                    || actionId == EditorInfo.IME_ACTION_DONE
                    || (event != null && KeyEvent.KEYCODE_ENTER == event.getKeyCode() && KeyEvent.ACTION_DOWN == event.getAction())) {
                login();
                return true;
            }
            return false;
        }
    }
}

File: app/src/main/java/com/zxcx/zhizhe/ui/loginAndRegister/login/LoginActivity.kt
package com.zxcx.zhizhe.ui.loginAndRegister.login

import android.os.Bundle
import android.os.CountDownTimer
import android.text.Html
import android.view.View
import butterknife.ButterKnife
import cn.jiguang.analytics.android.api.JAnalyticsInterface
import cn.jpush.android.api.JPushInterface
import cn.sharesdk.framework.Platform
import cn.sharesdk.framework.PlatformActionListener
import cn.sharesdk.framework.ShareSDK
import cn.sharesdk.sina.weibo.SinaWeibo
import cn.sharesdk.tencent.qq.QQ
import cn.sharesdk.wechat.friends.Wechat
import cn.sharesdk.wechat.utils.WechatClientNotExistException
import cn.smssdk.EventHandler
import cn.smssdk.SMSSDK
import com.google.gson.JsonParser
import com.jakewharton.rxbinding2.view.RxView
import com.meituan.android.walle.WalleChannelReader
import com.zxcx.zhizhe.R
import com.zxcx.zhizhe.event.LoginEvent
import com.zxcx.zhizhe.event.PhoneConfirmEvent
import com.zxcx.zhizhe.mvpBase.MvpActivity
import com.zxcx.zhizhe.ui.loginAndRegister.channelRegister.ChannelRegisterActivity
import com.zxcx.zhizhe.ui.my.selectAttention.SelectAttentionActivity
import com.zxcx.zhizhe.ui.welcome.WebViewActivity
import com.zxcx.zhizhe.utils.*
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import kotlinx.android.synthetic.main.activity_login.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import java.util.*
import java.util.concurrent.TimeUnit
import java.util.regex.Pattern

/**
 * 登录页面
 */

class LoginActivity : MvpActivity<LoginPresenter>(), LoginContract.View {

	private var phoneRules = "^1\\d{10}$"
	private var phonePattern = Pattern.compile(phoneRules)

	private var mChannelLoginListener: PlatformActionListener = ChannelLoginListener()
	private var channelType: Int = 0 // 1-QQ 2-WeChat 3-Weibo
	private var appType: Int = 0
	private var userName: String? = null
	private var userId: String? = null
	private var userIcon: String? = null
	private var userGender: String? = null
	private var appChannel: String? = null
	private var appVersion: String? = null
	private var jpushID: String? = null
	private var verifyKey: String? = null

	val countDownTimer: CountDownTimer = object : CountDownTimer(60000, 1000) {
		override fun onFinish() {
			tv_login_resend_code.isEnabled = true
			tv_login_resend_code.text = "重新获取验证码"
		}

		override fun onTick(millisUntilFinished: Long) {
			val value = (millisUntilFinished / 1000).toInt()
			tv_login_resend_code.text = getString(R.string.tv_login_count_down, value)
		}
	}

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_login)
		val uiFlags = View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or
				View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION or
				View.SYSTEM_UI_FLAG_FULLSCREEN or
				View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
		window.decorView.systemUiVisibility = uiFlags
		ButterKnife.bind(this)
		EventBus.getDefault().register(this)


		val str = "注册智者代表你已阅读并同意<font color='#0088AA'>《智者用户协议》</font>";
		tv_login_agreement.setText(Html.fromHtml(str));

		appType = Constants.APP_TYPE
		appChannel = WalleChannelReader.getChannel(mActivity)
		if (appChannel == null) {
			appChannel = "官方"
		}
		appVersion = Utils.getAppVersionName(mActivity)
		jpushID = JPushInterface.getRegistrationID(mActivity)
		SMSSDK.registerEventHandler(EventHandle())
	}

	override fun initStatusBar() {
		//全屏输入法问题
		AndroidBug5497Workaround.assistActivity(this)
	}

	override fun onBackPressed() {
		Utils.closeInputMethod(this)
		super.onBackPressed()
		countDownTimer.cancel()
	}

	public override fun onDestroy() {
		SMSSDK.unregisterAllEventHandler()
		EventBus.getDefault().unregister(this)
		super.onDestroy()
	}

	override fun createPresenter(): LoginPresenter {
		return LoginPresenter(this)
	}

	override fun getDataSuccess(bean: LoginBean) {
		if (!bean.user.hasAttention) {
//			SharedPreferencesUtil.saveData(SVTSConstants.isFirstLogin, false)
			startActivity(SelectAttentionActivity::class.java) {}
		}
		ZhiZheUtils.saveLoginData(bean)
		//极光统计
		val lEvent = cn.jiguang.analytics.android.api.LoginEvent("defult", true)
		lEvent.addKeyValue("appChannel", WalleChannelReader.getChannel(mActivity)).addKeyValue("appVersion", Utils.getAppVersionName(mActivity))
		JAnalyticsInterface.onEvent(mActivity, lEvent)
		//登录成功通知
		EventBus.getDefault().post(LoginEvent())
		Utils.closeInputMethod(mActivity)
		toastShow("欢迎来到智者")
		finish()
	}

	override fun channelLoginSuccess(bean: LoginBean) {
		getDataSuccess(bean)
	}

	override fun channelLoginNeedRegister() {
		mActivity.startActivity(ChannelRegisterActivity::class.java) {
			it.putExtra("userId", userId)
			it.putExtra("userName", userName)
			it.putExtra("userIcon", userIcon)
			it.putExtra("userGender", userGender)
			it.putExtra("channelType", channelType)
		}
		finish()
	}

	@Subscribe(threadMode = ThreadMode.MAIN)
	fun onMessageEvent(event: PhoneConfirmEvent) {
		//手机号确认成功,发送验证码
		showLoading()
		SMSSDK.getVerificationCode("86", et_login_phone.text.toString())
	}

	private fun checkPhone(): Boolean {
		return phonePattern.matcher(et_login_phone.text.toString()).matches()
	}

	override fun setListener() {
		iv_login_close.setOnClickListener { onBackPressed() }

		tv_login_phone.setOnClickListener {
			rl_login_phone.visibility = View.VISIBLE
			ll_login_code.visibility = View.GONE
			Observable.timer(300, TimeUnit.MILLISECONDS, AndroidSchedulers.mainThread()).subscribe {
				et_login_phone.requestFocus()
			}
		}

		tv_login_send_code.setOnClickListener {
			val confirmDialog = PhoneConfirmDialog()
			val bundle = Bundle()
			bundle.putString("phone", et_login_phone.text.toString())
			confirmDialog.arguments = bundle
			confirmDialog.show(mActivity.supportFragmentManager, "")
		}

		tv_login_resend_code.setOnClickListener {
			SMSSDK.getVerificationCode("86", et_login_phone.text.toString())
		}

		iv_login_phone_clear.setOnClickListener {
			et_login_phone.setText("")
			Utils.showInputMethod(et_login_phone)
		}

		RxView.clicks(iv_login_qq).throttleFirst(2000, TimeUnit.MILLISECONDS).subscribe {
			channelType = 1
			val qq = ShareSDK.getPlatform(QQ.NAME)
			qq.platformActionListener = mChannelLoginListener
			qq.showUser(null)
		}

		RxView.clicks(iv_login_wechat).throttleFirst(2000, TimeUnit.MILLISECONDS).subscribe {
			channelType = 2
			val wechat = ShareSDK.getPlatform(Wechat.NAME)
			wechat.platformActionListener = mChannelLoginListener
			wechat.showUser(null)
		}

		RxView.clicks(iv_login_weibo).throttleFirst(2000, TimeUnit.MILLISECONDS).subscribe {
			channelType = 3
			val weibo = ShareSDK.getPlatform(SinaWeibo.NAME)
			weibo.platformActionListener = mChannelLoginListener
			weibo.showUser(null)
		}

		et_login_phone.afterTextChanged {
			val isPhone = checkPhone()
			if (isPhone) {
				Utils.closeInputMethod(et_login_phone)
				et_login_phone.clearFocus()
			}
			iv_login_phone_clear.visibility = if (isPhone || et_login_phone.length() == 0) View.GONE else View.VISIBLE
			tv_login_send_code.isEnabled = isPhone
		}

		vci_login.setOnCompleteListener {
			Utils.closeInputMethod(et_login_phone)
			mPresenter.smsCodeLogin(et_login_phone.text.toString(), it, jpushID, appType, appChannel, appVersion)
		}

		tv_login_agreement.setOnClickListener {
			startActivity(WebViewActivity::class.java) {
				it.putExtra("title", getString(R.string.agreement))
				if (Constants.IS_NIGHT) {
					it.putExtra("url", getString(R.string.base_url) + getString(R.string.agreement_dark_url))
				} else {
					it.putExtra("url", getString(R.string.base_url) + getString(R.string.agreement_url))
				}
			}
		}
	}

	internal inner class ChannelLoginListener : PlatformActionListener {

		override fun onComplete(platform: Platform, i: Int, hashMap: HashMap<String, Any>) {
			if (i == Platform.ACTION_USER_INFOR) {
				val platDB = platform.db//获取数平台数据DB
				//通过DB获取各种数据
				userGender = platDB.userGender
				userIcon = platDB.userIcon
				userId = platDB.userId
				userName = platDB.userName

                val Uid = platDB.get("unionid")
                LogCat.e("UID" + Uid)
                LogCat.e("登录前ID"+userId)
				mPresenter.channelLogin(channelType, Uid, jpushID, appType, appChannel, appVersion)
			}
		}

		override fun onError(platform: Platform, i: Int, throwable: Throwable) {
			throwable.printStackTrace()
			mActivity.runOnUiThread {
				if (throwable is WechatClientNotExistException) {
					toastShow("请先安装微信客户端")
				} else {
					toastShow("登录失败")
				}
			}
		}

		override fun onCancel(platform: Platform, i: Int) {
			mActivity.runOnUiThread { toastShow("登录取消") }
		}
	}

	internal inner class EventHandle : EventHandler() {
		override fun afterEvent(event: Int, result: Int, data: Any) {
			mActivity.runOnUiThread {
				hideLoading()
				if (result == SMSSDK.RESULT_COMPLETE) {
					//回调完成
					when (event) {
						SMSSDK.EVENT_GET_VERIFICATION_CODE -> {
							//获取验证码成功
							toastShow(R.string.send_verification_code_success)
							tv_login_phone.text = et_login_phone.text.toString()
							rl_login_phone.visibility = View.GONE
							ll_login_code.visibility = View.VISIBLE
							tv_login_resend_code.isEnabled = false
							Observable.timer(300, TimeUnit.MILLISECONDS, AndroidSchedulers.mainThread()).subscribe {
								vci_login.focus()
							}
							countDownTimer.start()
						}
					}
				} else {
					// 根据服务器返回的网络错误，给toast提示
					try {
						val throwable = data as Throwable
						throwable.printStackTrace()
						val `object` = JsonParser().parse(throwable.message).asJsonObject
						val des = `object`.get("detail").asString//错误描述
						val status = `object`.get("status").asInt//错误代码
						when (status) {
							457 -> toastError("手机号格式错误!")
							463, 464, 465, 477, 478 -> toastError("获取验证码次数频繁，请稍后重试")
							else -> toastError(des)
						}
					} catch (e: Exception) {
						//do something
					}

				}
			}
		}
	}
}
